lisa/util/shell.py

```diff
@@ -380,6 +391,7 @@ def spawn(
                 # Except CommandInitializationError then use minimal shell type.
                 if not have_tried_minimal_type:
                     self._inner_shell._spur._shell_type = spur.ssh.ShellTypes.minimal
+                    _minimize_shell(self._inner_shell)
```

Comment:
Hi @lubaihua33. I'm sorry you see timeouts there, but that needs further debugging. There is nothing in setting the generate_run_command() function on the shell to a simple, continuous string with the individual tokens (no added quotes) that would make a Linux shell unhappy, I'm afraid. But there is definitely a reason non-POSIX shells would reject that, and we need the support.

The regexp will match regardless, as it looks for a pattern added by spur: https://github.com/mwilliamson/spur.py/blob/master/spur/errors.py#L11, so this is not really helping in my case.

I will fix the style errors, but this change is not really sound. I'd kindly ask for your debugging as to why it fails on your side, so we get to a better fix for that case.

Thanks.

Reply:
**Some background:**
The images whose publishers are "azhpc-desktop", "javlinltd" are the POSIX shells. But calling sqawn to run any Linux commands might raise CommandInitializationError. Using minimal shel type can solve this. I added comments here for the reason https://github.com/microsoft/lisa/blob/main/lisa/util/shell.py#L370.

**The reason that fails on my side:**
Overriding the generate_run_command() with the minimal_generate_run_command will remove some quotes for commands. However, removing some quotes for a command might lead to a timeout for a POSIX shell.

For example, 
The return value of the original fun 'generate_run_command' is
`''sudo' 'sh' '-c' 'cat /etc/shadow | grep -v "lisatest"''` (This is good for a POSIX shell)

The return value of the override one minimal_generate_run_command is
`'sudo sh -c cat /etc/shadow | grep -v "lisatest"`'  (There are no quotes for `cat /etc/shadow | grep -v "lisatest"`. If running "sudo sh -c cat /etc/shadow" without quotes for "cat /etc/shadow" in a POSIX shell, only the first string "cat" will be the value of "-c". And only running "cat" will lead to a timeout in a POSIX shell.

Those systems of "azhpc-desktop" and "javlinltd" images are POSIX shells and will enter CommandInitializationError exception. Running some commands like the above ones will timeout.

That's why it fails (timeout) for the systems of "azhpc-desktop", "javlinltd" images.

**How to fix**
I have no better way to find a middle ground for both your scenario and my scenario. Just found the system of "azhpc-desktop", "javlinltd" images of my scenario can match _spawn_initialization_error_pattern, however the system of your scenario won't match. So I thought if we can use this to distinguish them. 
This change is not really sound. That would be great if you could find a better way.

Reply:
I see. I guess I was lazy/greedy in the substitutions there, my bad. What if we only cut off the obviously shlex.split() added quotes there, like on this latest update? I think it will make your case behave again. Ty

Reply:
I still got timeout. You can run "verify_no_pre_exist_users" test case for image "javlinltd1579185328273 cloverdx-server-vm-image 6-5-0 1.0.0" to reproduce.

Reply:
Would you please give me a full cmdline for that? I have very little experience with running upstream lisa alone...

Reply:
Let me send you the runbook and the command offline.

Reply:
Just did it and it passes with my last change on the patch:

```
2024-08-19 16:51:31.450[20688][INFO] lisa.RootRunner ________________________________________
2024-08-19 16:51:31.450[20688][INFO] lisa.RootRunner       AzureImageStandard.verify_no_pre_exist_users: PASSED
2024-08-19 16:51:31.450[20688][INFO] lisa.RootRunner test result summary
2024-08-19 16:51:31.450[20688][INFO] lisa.RootRunner     TOTAL    : 1
2024-08-19 16:51:31.450[20688][INFO] lisa.RootRunner     QUEUED   : 0
2024-08-19 16:51:31.450[20688][INFO] lisa.RootRunner     ASSIGNED : 0
2024-08-19 16:51:31.450[20688][INFO] lisa.RootRunner     RUNNING  : 0
2024-08-19 16:51:31.450[20688][INFO] lisa.RootRunner     FAILED   : 0
2024-08-19 16:51:31.450[20688][INFO] lisa.RootRunner     PASSED   : 1
2024-08-19 16:51:31.450[20688][INFO] lisa.RootRunner     SKIPPED  : 0
2024-08-19 16:51:31.450[8828][DEBUG] lisa.notifier[Console] received message [TestRun]: TestRunMessage(type='TestRun', time=datetime.datetime(2024, 8, 19, 16, 51, 31, 450097), elapsed=193.20371299999533, status=<TestRunStatus.SUCCESS: 3>, runbook_name=None, test_project=None, test_pass=None, tags=None, run_name='lisa-None-20240819-164815-879', message='')
2024-08-19 16:51:31.450[20688][INFO] lisa. completed in 195.570 sec
```

Reply:
I run this test on your branch again. It still failed. I have sent you the log hope we can find the gap.

Reply:
My bad, too many copies of the repo led to a slight confusion/mismatch. OK, this is a version that works for all. escape_sh() from spur was doing basically what I do here now. But that function was not just escaping quotes, it was adding quotes around things too. Now I just do that when spaces are found in the input, in the minimal case.

Reply:
Thanks for this change. It works.